File: Array/easy/Missing_elements_in_an_sorted_array.cpp
//
// Created by santosh on 12/1/25.
//
#include<bits/stdc++.h>
using namespace std;

vector<int> missing_elements_method1(vector<int> nums)
{
    vector<int> result;
    for (int i = 0; i < nums.size() - 1; i++)
    {
        if (nums[i + 1] - nums[i] != 1)
        {
            int j = nums[i] + 1;
            while (j < nums[i + 1])
            {
                result.push_back(j);
                j++;
            }
        }
    }

    return result;
}

vector<int> missing_elements_method2(vector<int> nums)
{
    vector<int> result;
    int max_el = *max_element(nums.begin(), nums.end());
    int min_el = *min_element(nums.begin(), nums.end());
    vector<int> hash_map(max_el + 1, 0); // all initialized to 0
    for (int i = 0; i < nums.size() - 1; i++)
    {
        hash_map[nums[i]]++;
    }

    for (int i = min_el; i < hash_map.size()-1; i++)
    {
        if (hash_map[i] == 0)
        {
            result.push_back(i);
        }
    }
    return result;
}

int main()
{
    vector<int> nums = {1, 2, 3, 6, 7, 8, 12, 24};
    vector<int> all_Missing_elements_from_method1 = missing_elements_method1(nums);
    cout << "Missing elements are method 1: ";
    for (int all_Missing_element : all_Missing_elements_from_method1)
    {
        cout << all_Missing_element << " ";
    }
    cout << endl;

    vector<int> all_Missing_elements_from_method2 = missing_elements_method2(nums);
    cout << "Missing elements are meethod 2 : ";
    for (int all_Missing_element : all_Missing_elements_from_method2)
    {
        cout << all_Missing_element << " ";
    }
}


// Time complexity - o(n+m)
// n - is all the elements in an array
// m - missing elements in an array
// worst case {0,10000) - o(m)

//Space Complexity:
//➡️ O(M)
// Where M is the number of missing elements.
//Except that, I only use a few extra variables → constant space.

File: Array/easy/Missing_elements_in_an_sorted_array.py
# Missing elements in a sorted array
# Created by santosh

def missing_elements_method1(nums: list[int]) -> list[int]:
    """
    Iterate through consecutive pairs and fill gaps.
    Time: O(n + m), Space: O(m)
    """
    result = []
    for i in range(len(nums) - 1):
        if nums[i + 1] - nums[i] != 1:
            j = nums[i] + 1
            while j < nums[i + 1]:
                result.append(j)
                j += 1
    return result


def missing_elements_method2(nums: list[int]) -> list[int]:
    """
    Uses a hash array to mark existing elements.
    Time: O(n + max_el), Space: O(max_el)
    """
    result = []
    max_el = max(nums)
    min_el = min(nums)
    hash_map = [0] * (max_el + 1)

    for num in nums:
        hash_map[num] += 1

    for i in range(min_el, len(hash_map) - 1):
        if hash_map[i] == 0:
            result.append(i)

    return result


# Time complexity - O(n + m)
# n - all elements in the array
# m - missing elements
# Space Complexity: O(m) where m is the number of missing elements


if __name__ == "__main__":
    nums = [1, 2, 3, 6, 7, 8, 12, 24]

    method1_result = missing_elements_method1(nums)
    print("Missing elements (method 1):", method1_result)

    method2_result = missing_elements_method2(nums)
    print("Missing elements (method 2):", method2_result)
